### ISSUE TITLE
gevent compatibility

### ISSUE THREAD

**Issue by zj8487** [grpc/grpc#4629]
Hi, I want to make grpc client and server code to work with gevent, is grpc compatible with gevent?

**Comment by vinays** [comment]
@zj8487 did you get gRPC(python) working with gevent.monkeypatch? Could you please share the work-around.

**Comment by zj8487** [comment]
@vinays 
it is sad, it is not work with gevent.
i do it as follow:
a thread(not monkey patched) to deal with the grpc and send or receive message from a message queue which was listen or loop in a gevent to process the request or response.

**Comment by mayflaver** [comment]
grpc make async io well， why do you want to make grpc client and server code to work with gevent? they do the same thing.

**Comment by vinays** [comment]
Thanks @zj8487 

@mayflaver in our case 'grpc' and 'gevent' are transitive dependency of gcloud-python and gunicorn respectively in our legacy app. If I was writing an app from scratch I would agree with you, but in this case these dependencies are not enforced by me. :(
To best of my knowledge--gunicorn with gevent as worker_class monkey patches a lot of standard library functions including threading. This conflicts with the use of "futures" in grpc and hangs the process.
Since, gunicorn(w/ gevent) is quite commonly used in the industry and anyone who wanna use gcloud-python(or other grpc based libs) in their app will likely run in to the same issue. I am open to any suggestions that might help solve this issue without changing the dependencies too much. Maybe gRPC and gunicorn teams should talk and figure out a solution?

**Comment by soltanmm** [comment]
Reopening as this seems to warrant investigating.

**Comment by vinays** [comment]
@soltanmm any resolution on this problem? It will be really help if we can get a rough timeline for it.

**Comment by soltanmm** [comment]
@vinays There's no timeline for it at present (and without in-depth `gevent` knowledge we're not exactly leaping into action on it), but as there appears to be mounting community pressure I guess I'll take a quick look tomorrow. Feel free to proactively provide a minimal reproduction; it might speed things along.

**Comment by soltanmm** [comment]
@vinays We have a hard time justifying interoperability with a library that seems to be [unable to work with standard library elements in Python 3](http://www.gevent.org/gevent.monkey.html#gevent.monkey.patch_thread) because it literally changes the standard library, and incompletely so.

That said, it's worth documenting this, and I don't know much about `gunicorn`. What's wrong with the other options [listed here](http://docs.gunicorn.org/en/stable/settings.html#worker-class) for your use case?

**Comment by soltanmm** [comment]
@vinays Actually, this issue comment might be of relevance to you: https://github.com/benoitc/gunicorn/issues/927#issuecomment-102176547

It appears that this problem arises with Django as well and what the `gunicorn` folks end up doing *might* be the fix you need.

**Comment by vinays** [comment]
Thanks @soltanmm for looking into the issue.

@mickgarvey @grourk might have more insight into why we did not choose other options in the [list](http://docs.gunicorn.org/en/stable/settings.html#worker-class). From what I know, since our app is IO bound 'sync' worker does not scale well because of blocking IO.

@soltanmm did your investigation conclude a root cause of the issue? Do you know what part of standard library patching was causing the hang in Python 2.7? Would excluding 'thread' from patch_all, fix the issue? I can test that myself, but I am looking for any side-effects you might know and most likely won't surface itself in a simple test. Any other information from your investigation might also help.

**Comment by soltanmm** [comment]
@vinays I'll have to guiltily admit that after reading about what `gevent` monkeypatching actually *is* and checking the docs on known hangs interacting with library elements we appear to use, I face-palmed and stopped before attempting a reproduction. I'll take a look again on Monday and get some first-hand results if that'll allow everyone to sleep a little better.

I have a strong feeling it's just going to be the usual, "Green-threads cannot be cooperative if they block on something that isn't cooperative," sort of deal, though (where 'something that isn't cooperative' could be anything the `gevent` folks missed while monkeypatching [which I have a feeling includes `concurrent.futures`] and every blocking gRPC core operation).

**Comment by soltanmm** [comment]
It's basically that. gRPC Python uses multiple threads for a server (e.g. one for the server receiving messages, one to pass the call data along to the application). These trip over each other when they don't cooperatively release polling to `gevent` (which at least one of them can't do because it's in gRPC core code and the core is obviously not based on `gevent`).

Short of getting in-depth knowledge of `gevent` (and I'm not a `gevent` dev, so, nope), I can't come up with much more than saying, "Excluding `thread` from monkey-patching seems to work."

I'm not going to close this as it's *possible* that there exists a way to force the use of standard Python threads despite monkey patching where they are necessary, or that some other hacky-solution-against-a-hacky-solution exists, but **we will not attempt to support gevent's monkeypatching** for the near term.

**Comment by nathanielmanistaatgoogle** [comment]
I'm still coming up to speed on what's being asked here, but from what I've gathered so far, I suspect it will be a lot: one of the core design assumptions of gRPC Python is that a reasonable number of threads will always be available to be devoted to blocking operations such as C extension calls and calls into application code. Rolling back that assumption and moving forward with a different structure are going to be substantial work, if they're something that we ever do at all.

I feel like I'm still missing quite a bit. What's the minimum sample program you could write that you would expect to work but that doesn't?

**Comment by vinays** [comment]
@soltanmm thanks for the explanation, it resonates with my understanding.
@nathanielmanistaatgoogle if your questions was pointed at me, I can write a sample script later today to reproduce the problem.

**Comment by nathanielmanistaatgoogle** [comment]
@vinays: yes, a code sample would be delightful - something that is enough to illustrate a use of gRPC Python, how you expected gRPC Python to behave, and how gRPC Python actually behaved.

**Comment by vinays** [comment]
@nathanielmanistaatgoogle here is some [sample code](https://github.com/vinays/grpc-gevent-mokey) to illustrate the issue. Let me know if you have any questions.

**Comment by Globegitter** [comment]
Just another use-case we have been running into. We are using graphql (https://github.com/graphql-python/graphene) as our public API and grpc for internal communication. 

The graphql library supports running resolving requests in parallel using gevent but the problem is, that any grpc calls we are making (in different resolve methods) are run serially, which now is the cause of quite a few performance issues.

I have just started to investigate, to find any potential solutions or workarounds, but good to see there are other people with similar issues.

**Comment by soltanmm** [comment]
@nathanielmanistaatgoogle Thoughts on gevent/gevent#786? We could leave an unsupported (private, but documented) hook for the underlying thread pool executor somewhere in the gRPC implementation and allow users to swap it out post-import/pre-absolutely-anything-else if they want to at their own risk.

**Comment by soltanmm** [comment]
Better yet: we could do that and attach warnings (like the deprecation warnings) to it in terminal emboldened firetruck-red all-caps font that the users have entered monkey territory and shouldn't be surprised if their hats get stolen and everything hurts twice as much.

**Comment by nathanielmanistaatgoogle** [comment]
Thank you @jamadden for [your investigation](https://github.com/gevent/gevent/issues/786#issuecomment-203988312).

Use of the same thread, with program control passing from application to gRPC Python to the gRPC Core and then to a blocking system call is an explicit part of gRPC’s design, is the basis for many of the performance improvements that we’ll be making in the next few months, and would be very hard to change. In particular major alterations would have to be made to the Core to enable its completion queues to be used in the manner gevent would want to use them. I’ve spoken with @ctiller about it and this work may be scheduled engineering effort if it is forecast to be valuable enough to be prioritized above other goals.

There may be a certain recipe for monkeypatching that happens to work today but we wouldn’t be able to offer proper support (with test coverage and a promise not to break in the future) for anything based around selectively patching thread pools used internally within gRPC Python. In particular we have near-term plans to scale back how widely we use thread pools. If you want today to do something based on selectively patching it may work fantastically for you and then break in a future release. I would celebrate your bravery and wish you fun.

**Comment by nathanielmanistaatgoogle** [comment]
Retitled this since the investigation is over.

Compatibility with `gevent` might happen, and if it ever does it will be following changes to the C Core.

**Comment by zj8487** [comment]
@nathanielmanistaatgoogle 
+1 for Compatibility with gevent.
one more sugesstion(may be not suit talk here.): really hope pure python version or non threads version.

grpc core with number of threads model that make it not convenient to work with the other frame like django uwsgi, etc. because it is hard to control the grpc client's life cycle. 

until now i am searching how to use grpc python client in django with uwsgi make it can support reload, and make the server can exit correctly.   :(

**Comment by nathanielmanistaatgoogle** [comment]
[A relevant thread on the `grpc.io` mailing list.](https://groups.google.com/d/topic/grpc-io/CzgL7yB4JdQ/discussion)

**Comment by public** [comment]
It is possible to get this working using gevent 1.2a (aka GitHub master.)

Use this `concurrent.futures` compatible ThreadPoolExecutor implementation from gevent http://www.gevent.org/gevent.threadpool.html#gevent.threadpool.ThreadPoolExecutor

And set `gevent.monkey.patch_all(thread=False)`.

At least it doesn't hang anymore when you do this :)

**Comment by renstrom** [comment]
If I understand the documentation correctly, you don't need to pass `thread=False`.

>A version of concurrent.futures.ThreadPoolExecutor that **always uses native threads, even when threading is monkey-patched**.

**Comment by nathanielmanistaatgoogle** [comment]
Just kicked off [a community process for addressing this enhancement request](https://groups.google.com/d/msg/grpc-io/RpkyqqQy8TU/UjL3ldKuAAAJ); please participate!

**Comment by srinathgs** [comment]
@nathanielmanistaatgoogle @soltanmm With gevent, 
```
File "/Users/srinath.gs/.virtualenvs/vajra/lib/python2.7/site-packages/grpc/_channel.py", line 670, in channel_spin
    event = state.completion_queue.poll()
  File "src/python/grpcio/grpc/_cython/_cygrpc/completion_queue.pyx.pxi", line 105, in grpc._cython.cygrpc.CompletionQueue.poll (src/python/grpcio/grpc/_cython/cygrpc.c:9612)
KeyboardInterrupt
```
I got a hang on a service that returns a stream as a response. The possible culprit is a busy while loop in the completion_queue.poll function.


`.proto` definition
```
  rpc Read(ReadFileRequest) returns (stream ReadFileResponse);
```
```
  def read(self, chunk=1024):
        rfreq = file_pb2.ReadFileRequest(file=self.f, chunk=chunk)
        for resp in self.stub.Read(rfreq, self.timeout):
            print resp
```

```
  print "reading"
  f.read()
```
execution does not get past the print message

**Comment by charyorde** [comment]
@nathanielmanistaatgoogle Thanks for the work. In what release version or milestone is this filed under please?

**Comment by nathanielmanistaatgoogle** [comment]
@charyorde: our current thinking is that we may be able to address it before the end of the calendar year.

**Comment by charyorde** [comment]
Thank you @nathanielmanistaatgoogle

**Comment by scorpioWalker** [comment]
gevent v1.2.2 with grpcio==1.4.0 hangs in RawGreenlet switch
I did not look into this issue, just rollback grpcio to v1.1.3 and it worked fine.
think soneone should know this.

**Comment by nicolasnoble** [comment]
@scorpioWalker: I think this is something you should create a new issue about. This one issue here is more integrating grpc with gevent. That's different from simply being able to run next to it.

**Comment by zero-master** [comment]
Google Datastore with Gunicorn blocks forever in app engine flexible. 
I am not sure if it's because of this issue.

gevent==1.2.2
google-api-core==0.1.1
google-auth==1.2.1
google-cloud-core==0.28.0
google-cloud-datastore==1.4.0
google-gax==0.15.16
google-resumable-media==0.3.1
googleapis-common-protos==1.5.3
greenlet==0.4.12
grpcio==1.7.0
gunicorn==19.7.1

**Comment by lita** [comment]
Hello! Would love to hear the status of this!

**Comment by AndreCimander** [comment]
We could use this as well, spawning processes is a waste of memory and cpu for heavy-I/O bound tasks...

**Comment by Revolution1** [comment]
any update?

**Comment by smetj** [comment]
hi @kpayson64 & devs,
I understand this issue is difficult or doesn't get any priority but would you pls communicate any concrete progress on this?
The issue is open for +2 years. It's totally fine if this doesn't get solved/delivered but if that's the case can we pls close this issue so ppl can move on to an alternative?

**Comment by codesuki** [comment]
@kpayson64 I saw in your repo that you are still working on this. Thanks! Does it seem to work?

**Comment by kpayson64** [comment]
@smetj 
@codesuki 
Sorry about the delay.  There were a few high priority issues the popped up that I needed to address, and I was on vacation for a week so this got pushed back a bit.

This is very much at the top of my stack right now.  I'm currently tackling some lingering test failures, but the bulk of this seems to be working.

**Comment by codesuki** [comment]
Sounds great! Thanks :) I hope we can get our gRPC working with locust soon.

**Comment by smetj** [comment]
Super excited this is moving forward! Thanks for the feedback ...
Can't wait starting to use Google cloud pubsub with Gevent + Python!

**Comment by ekhaydarov** [comment]
is aiohttp a viable alternative?

**Comment by lita** [comment]
Same! I am excited to try this out.

**Comment by raybotha** [comment]
Would be great to get https://github.com/grpc/grpc/pull/14561 merged in

**Comment by kpayson64** [comment]
With #14561 in, you should now be able to use gevent.

Note that you have to explicitly monkey patch Python, as well as enabling the gevent loop for Python (Before doing anything with gRPC)

```
from gevent import monkey
monkey.patch_all()

import grpc.experimental.gevent as grpc_gevent
grpc_gevent.init_gevent()
```
This will be included in the 1.11 release.

I'm closing this master issue, and separate issues can be created with any issues encountered using gevent.

**Comment by mitar** [comment]
For those who want asyncio support, can #7910 be reopened or should we open a new issue?

**Comment by mitar** [comment]
Or is #6046 that?

**Comment by charyorde** [comment]
Finally. Testing away!

**Comment by codesuki** [comment]
@kpayson64 thanks, great job!

**Comment by kpayson64** [comment]
@mitar 
Feel free to open a new issue.  I'd expect that we need to add some new surface APIs for asyncio support, so its probably a good idea to have a separate issue.

**Comment by mitar** [comment]
I opened #14876. Based on existing gist (linked in the issue) it seems this could be easily done.

**Comment by CarolineChristie** [comment]
The official documentation needs to be corrected. It links to this issue and says:
> Warning: Because using the gevent worker is known to cause issues with gRPC, we recommend that you do NOT use gevent workers if you're using the Google Cloud Client Library, which uses gRPC. Using gevent workers with the Google Cloud Client Library will cause requests to hang.

https://cloud.google.com/appengine/docs/flexible/python/runtime
